### PR TITLE
Refactor layout implementation

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
Cleaning up some layout type related arguments, previously we are putting layout specific args like `inner_k_tiles` in `from_float`, but now we just use `layout_type` which can be different `LayoutType` instances and can hold different args for each type of LayoutType

Test Plan:
regression tests: 
```
python test/quantization/test_quant_api.py
python test/integration/test_integration.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torchao/dtypes/utils.py [hunk 1]
```diff
@@ -29,6 +30,12 @@ def wrapper(*args, **kwargs):
     return decorator
 
 """
+Layout type definition
+"""
+@dataclass
```
Comment (msaroufim) [inline torchao/dtypes/utils.py:35]: Add a comment that this shouldn't be instantiated directly

File: torchao/dtypes/utils.py [hunk 2]
```diff
@@ -37,29 +44,29 @@ def wrapper(*args, **kwargs):
 """
 _LAYOUT_CONSTRUCTOR_TABLE: Dict[Callable, Dict[str, Callable]] = defaultdict(dict)
 
-def _register_layout_cls(cls: Callable, extended_layout: str):
+def _register_layout_cls(cls: Callable, layout_name: str):
```
Comment (msaroufim) [inline torchao/dtypes/utils.py:47]: I thought the main benefit of this PR was no longer having to use layout_name or stringly interfaces at all

Comment (jerryzh168) [reply torchao/dtypes/utils.py:47]: sure I can replace this with the type (class) for LayoutType I think

File: torchao/dtypes/affine_quantized_tensor.py [hunk 3]
```diff
@@ -194,13 +203,10 @@ def from_float(
...
         zero_point_domain: ZeroPointDomain = ZeroPointDomain.INT,
-        extended_layout: str = "plain",
-        # TODO: this is only for "tensor_core_tiled", need to figure out
-        # the proper API for this arg
-        inner_k_tiles: Optional[int] = None,
+        layout_type: LayoutType = PlainLayoutType(),
     ):
         original_shape = input_float.shape
-        if extended_layout == "tensor_core_tiled":
+        if isinstance(layout_type, TensorCoreTiledLayoutType):
```
Comment (vayuda) [inline torchao/dtypes/affine_quantized_tensor.py:209]: is it possible to move this inside the TensorCoreTiledLayout constructor

Comment (jerryzh168) [reply torchao/dtypes/affine_quantized_tensor.py:209]: oh, this is padding the input though, not sure how we can do that, but we can move the implementation to `TensorCoreTiledLayout` I think

Comment (vayuda) [reply torchao/dtypes/affine_quantized_tensor.py:209]: Yea I think that would be a bit cleaner if all layout specific code was kept within the layout class.

**Review by vayuda — APPROVED** [review]

File: torchao/dtypes/affine_quantized_tensor.py [hunk 4]
```diff
@@ -20,10 +20,32 @@
...
     _get_layout_tensor_constructor,
+    LayoutType,
 )
+from typing import ClassVar
+from dataclasses import dataclass
 
 aten = torch.ops.aten
 
+@dataclass(frozen=True)
+class PlainLayoutType(LayoutType):
```
Comment (msaroufim) [inline torchao/dtypes/affine_quantized_tensor.py:31]: comment or error that this shouldnt be instantiated directly

File: torchao/dtypes/affine_quantized_tensor.py [hunk 5]
```diff
@@ -20,10 +20,32 @@
...
+    pass
+
+@dataclass(frozen=True)
+class TensorCoreTiledLayoutType(LayoutType):
+    inner_k_tiles: int = 8
+
+    def pad_input(self, input: torch.Tensor) -> torch.Tensor:
+        orig_out_features, orig_in_features = input.shape
+        in_features = find_multiple(orig_in_features, 1024)
```
Comment (msaroufim) [inline torchao/dtypes/affine_quantized_tensor.py:40]: where are the in and out numbers coming from? I constants like this were a function of the dtype as well

File: torchao/dtypes/affine_quantized_tensor.py [hunk 6]
```diff
@@ -20,10 +20,32 @@
...
     _register_layout_cls,
     _get_layout_tensor_constructor,
+    LayoutType,
 )
+from typing import ClassVar
+from dataclasses import dataclass
 
 aten = torch.ops.aten
 
+@dataclass(frozen=True)
```
Comment (msaroufim) [inline torchao/dtypes/affine_quantized_tensor.py:30]: not sure what benefit we're getting from using dataclasses, what you want seems to be closer to an enum with 2 values Plain or TensorCoreTiled and specifically if using TensorCoreTiled there's a util function called pad_input that doesn't have to be part of the class

File: torchao/dtypes/affine_quantized_tensor.py (hunk 4 above)
Comment (jerryzh168) [reply torchao/dtypes/affine_quantized_tensor.py:31]: this can be instantiated I think, are you talking about LayoutType?

File: torchao/dtypes/utils.py [hunk 7]
```diff
@@ -28,38 +30,53 @@ def wrapper(*args, **kwargs):
         return func
     return decorator
 
+"""
+Base class for different LayoutType, should not be instantiated directly
+"""
+@dataclass(frozen=True)
+class LayoutType:
+    def pad_input(self, input: torch.Tensor) -> torch.Tensor:
```
Comment (msaroufim) [inline torchao/dtypes/utils.py:38]: This is kind of misleading and should return a not implemented error in my opinion

The better way to do this is to add   `@abstractmethod` after importing `from abc import ABC, abstractmethod` so the error gets raised correctly

File: torchao/dtypes/affine_quantized_tensor.py (hunk 5 above)
Comment (jerryzh168) [reply torchao/dtypes/affine_quantized_tensor.py:40]: this is comes from tinygemm kernel I think, this layout only applies to int4 dtype

File: torchao/dtypes/utils.py (hunk 7 above)
Comment (jerryzh168) [reply torchao/dtypes/utils.py:38]: not all layout needs padding though, why do we want to raise an error here?

File: torchao/dtypes/affine_quantized_tensor.py (hunk 4 above)
Comment (msaroufim) [reply torchao/dtypes/affine_quantized_tensor.py:31]: I see I guess I'm a bit thrown off because a data classes primary goal is to store data wheras this class stores nothing

Comment (msaroufim) [reply torchao/dtypes/affine_quantized_tensor.py:31]: I would have instead done an enum like this

```
from enum import Enum
class Operations(Enum):
    ADD = (1,)
    SUBTRACT = (2,)
    MULTIPLY = (3,)
    DIVIDE = (4, 'precision') 
```

enums are also a class so you can override `__init__` and define a `func` that only applies on DIVIDE for example

Comment (jerryzh168) [reply torchao/dtypes/affine_quantized_tensor.py:31]: sorry I don't follow the `DIVIDE` part, can you elaborate a bit? is this talking about how to support `TensorCoreTiledLayoutType` that has a inner_k_tiles argument?

File: torchao/dtypes/affine_quantized_tensor.py [hunk 8]
```diff
@@ -20,10 +20,32 @@
...
+class PlainLayoutType(LayoutType):
+    pass
+
+@dataclass(frozen=True)
+class TensorCoreTiledLayoutType(LayoutType):
+    inner_k_tiles: int = 8
```
Comment (jerryzh168) [inline torchao/dtypes/affine_quantized_tensor.py:36]: @msaroufim see here, we have extra configurable arguments, it's not just a name so I'm not sure how enum would work here

**Review by msaroufim — APPROVED** [review]

**Review by msaroufim — APPROVED** [review]

File: torchao/dtypes/affine_quantized_tensor.py [hunk 9]
```diff
@@ -194,30 +214,16 @@ def from_float(
...
         original_shape = input_float.shape
-        if extended_layout == "tensor_core_tiled":
-            orig_out_features, orig_in_features = input_float.shape
-            in_features = find_multiple(orig_in_features, 1024)
-            out_features = find_multiple(orig_out_features, 8)
-            input_float = torch.nn.functional.pad(
-                input_float,
-                (0, in_features - orig_in_features, 0, out_features - orig_out_features),
-            )
+        input_float = layout_type.pad_input(input_float)
```
Comment (vayuda) [inline torchao/dtypes/affine_quantized_tensor.py:220]: This is on the right track but I think we can make it more generic to better serve future usage. 
LayoutTypes can implement two functions: `pre_process()` and  `post_process()` The new workflow would look like this:
```
input_float = layout.pre_process(input_float)
...
int_data = quantize_affine(...)
int_data = layout.post_process(int_data)
layout_tensor_ctr = get_layout_tensor_constructor(type(layout_type))
layout_tensor = layout_tensor_ctr(int_data, scale, zero_point, layout_type)
```
To motivate this I can say that I would for sure use the post_process function while integrating my intx work. This allows me to write just a layout_type instead of an entire layout_class. Additionally, if you look at  #498 they call torch._cslt_compress(int_data) which could also be implemented as a post_process function.

Comment (jerryzh168) [reply torchao/dtypes/affine_quantized_tensor.py:220]: makes sense, by reusing PlainLayoutType, you mean inherit it and override `post_process` right?

**Review by vayuda — APPROVED** [review]
